### PR TITLE
Button: Add opt-in prop for larger `isSmall` size

### PR DESCRIPTION
**Pull request by mirka**
Part of #46741

## What?

Adds a `__next32pxSmallSize` prop to opt into the new 32px sizes when `isSmall=true`.

## Why?

To ensure a smooth transition from the current small size (24px) to the new small size (32px).

## Testing Instructions

1. `npm run storybook:dev`
2. For the "Primary" and "Icon" stories for Button, toggle the control for the `__next32pxSmallSize` prop. This should have no effect when `isSmall=false`.

## Screenshots or screencast 

Top row is `isSmall`, bottom row is `isSmall` & `__next32pxSmallSize`.

The isSmall buttons in 24px and 32px sizes

### CODE REVIEW

File: packages/base-styles/_variables.scss [hunk 1]
```diff
@@ -45,6 +45,7 @@ $grid-unit-80: 8 * $grid-unit;		// 64px
  */
 
 $icon-size: 24px;
+$icon-size-next-default-32px: 32px; // transitionary variable for next small button size
```
Comment (ciampo) [inline packages/base-styles/_variables.scss]: It looks like `$icon-size` is used in a lot of places, and often for icons — I'm not sure that we should change this variable.

What about the `$button-size-small` variable a few lines below? Should we not be using that variable instead, also in the `Button` component?

**Review by ciampo — COMMENTED** [review]
I left a comment about using and tweaking the `$icon-size` variable — otherwise the fallback prop logic LGTM, and Storybook is working as expected.

File: packages/base-styles/_variables.scss (hunk 1 above)
Comment (mirka) [reply packages/base-styles/_variables.scss]: You're absolutely right, my bad 🙈 Thanks for noticing!

File: packages/components/src/button/style.scss [hunk 2]
```diff
@@ -263,15 +263,24 @@
...
-		height: $icon-size;
+		height: $button-size-small-next-default-32px;
 		line-height: 22px;
 		padding: 0 8px;
 		font-size: 11px;
 
 		&.has-icon:not(.has-text) {
 			padding: 0;
-			width: $icon-size;
-			min-width: $icon-size;
```
Comment (mirka) [inline packages/components/src/button/style.scss:274]: As @ciampo noted, these `$icon-size` variables should've been `$button-size-small` (both 24px though).

This error has also been fixed in this PR.

**Review by ciampo — APPROVED** [review]
LGTM 🚀 

Code changes look good, Storybook examples work as expected, and VizReg tests are passing.

**Comment by jasmussen** [comment]
Just catching up here, this is a great step towards https://github.com/WordPress/gutenberg/issues/46734, thank you.

I tried it today on the Publish button, and it looks like this:

Screenshot 2023-06-02 at 09 11 10

That's a small font, which isn't according ot the mockups, which simply changes the height of the button, note the toolbar buttons here:

**Comment by ciampo** [comment]
Thank you for the feedback, Joen! I guess we always assumed that https://github.com/WordPress/gutenberg/issues/46734 was about the components' height, and therefore ignored other metrics.

The small variant's font size is [currently hardcoded to be `11px`](https://github.com/WordPress/gutenberg/blob/19bcabfcbbc483028d1a429d9abe6a0fd595d233/packages/components/src/button/style.scss#L269), and it looks like it has been like that for 5+ years (#6562).

Should we update `Button`'s styles and have the small variant use the same font size as the default variant?

**Comment by jasmussen** [comment]
> Should we update Button's styles and have the small variant use the same font size as the default variant?

That's the big question. It seems like there are two paths forward:

1. Keep the "isSmall" version with a small font size, and create a new prop for the 32px non-default button size.
2. Replace the "isSmall" version so it has the same font size, height, and padding metrics as suggested on this issue.

I'm drawn to 2, because I don't understand the original context for why an isSmall exists, or where it's used. But that context is likely crucial in order to make a decision. Can we find out how widely the current isSmall is used?

For what it's worth, we could potentially have an isCompact for the 32px full-font-size version, and keep isSmall at 24px height, which would match tiny toggles like these:

Screenshot 2023-06-06 at 09 38 25

**Comment by ciampo** [comment]
I also am not aware of the context behind the specific need for the `isSmall` prop. Given that it was introduced more than 5 years ago, its original purpose may not hold true anymore and its usage may have evolved.

From a very simple and quick [initial search](https://github.com/search?q=isSmall+repo%3Awordpress%2Fgutenberg&type=code&ref=advsearch), it looks like it's used in a few places across the editor. Although I assume that its usage may be even more widespread across third party plugins/themes/apps, given how long this component has been around.

Personally, I'd avoid introducing more ad-hoc props / variants to the component unless absolutely necessary — and therefore I agree with your preference for option no. 2.

@mirka should be able to take a deeper look at this next week, after she's back from WCEU.

**Comment by jasmussen** [comment]
Sounds good. I think it's probably used for legacy "Reset" buttons. But another argument for going with approach 2 is that those were never that well-designed, and 11px is really illegible.

**Comment by mirka** [comment]
I split the `isSmall` problem into a new issue so we can discuss further 👉 #51708

If we want to keep both 24px and 32px sizes in the design system, we will likely need to introduce a `size` prop rather than continue with the `isSmall` prop.

**Pull request merged**